### PR TITLE
fix(datasource/maven): look for maven snapshot pom

### PR DESCRIPTION
**Pull request by bpfoster**
Maven snapshot poms are stored in the repository with timestamp & build number replacing the -SNAPSHOT in the name.

The check for missing artifacts did not resolve these files and so SNAPSHOTS always appeared missing and were filtered out.

Not sure if the async handling in `filterMissingArtifacts()` is the best, please let me know if changes needed.

## Changes:

Maven snapshot dependencies should now be included without wholesale disabling the pom check via `RENOVATE_EXPERIMENTAL_NO_MAVEN_POM_CHECK`.

## Context:

This should fix issue #3728

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by viceice — DISMISSED** [review]
@zharinov Can you please review this?

**Review by zharinov — CHANGES_REQUESTED** [review]

File: lib/datasource/maven/index.ts [hunk 1]
```diff
@@ -106,15 +166,17 @@ async function filterMissingArtifacts(
...
           dependency,
           repoUrl,
-          `${version}/${dependency.name}-${version}.pom`
+          await createUrlForDependencyPom(version, dependency, repoUrl)
         );
         return [version, artifactUrl];
-      })
+      }
+    );
+    const queue = (await Promise.all(versionUrls))
```
[the remark this answers is not in the document]
Comment (bpfoster) [reply lib/datasource/maven/index.ts]: Had a feeling this wasn't the best way to do this.  Apologies, my javascript knowledge is low.  Rewrote to use `pMap()` with no more `Promise.all()` - does that look better now?

File: lib/datasource/maven/index.spec.ts [hunk 2]
```diff
@@ -306,6 +338,72 @@ describe('datasource/maven/index', () => {
...
+          version: '1.0.1-SNAPSHOT',
+          meta: loadFixture('metadata-snapshot-version-invalid.xml'),
+        },
+      ],
+    });
+
+    httpMock
+      .scope(baseUrl)
+      .head('/org/example/package/1.0.1-SNAPSHOT/package-1.0.1-SNAPSHOT.pom')
+      .reply(404, '');
```
[the remark this answers is not in the document]
Comment (bpfoster) [reply lib/datasource/maven/index.spec.ts]: Rolled the snapshots into `mockGenericPackage`, is this what you were thinking?

**Review by rarkins — CHANGES_REQUESTED** [review]
Can you add some commenting to the primary logic to explain what's going on? This check logic has now got pretty large and would be increasingly challenging for anybody to debug or extend further

**Comment by bpfoster** [comment]
Peppered some comments in to the affected logic.

**Comment by rarkins** [comment]
Thanks @bpfoster. @zharinov can you review?

**Review by zharinov — APPROVED** [review]
Looks good to me.
@bpfoster, thanks for the PR of such a good quality!

**Review by rarkins — APPROVED** [review]

**Pull request merged**